portal-web/test/functional/com/liferay/portalweb/macros/LiferayEvent.macro

```diff
@@ -75,4 +75,31 @@ window.LFR_JS_EVENTS_LOG = {
 		RunScript(locator1 = "${javascript}");
 	}
 
+	macro waitForLiferayEvent {
+		var javascript = '''
+try {
+	var eventValue = LFR_JS_EVENTS_LOG.getLastEvent("${eventName}").${attributeName};
+
+	if (eventValue.includes("${attributeValue}")) {
+		return "fired";
```

Comment:
This API makes us lose so much information about the event that was "fired"... Is this what you where referring to when you said `getEval` only returned Strings?

Why not something like:

```javascript
const event = LFR_JS_EVENTS_LOG.getLastEvent("${eventName}");

// It would be better to use the [] syntax to protect from attribute names with illegal characters for the dot notation.
if (event['${attributeName}'] == '${attributeValue}') {
    return JSON.stringify(event);
}
```

Then:

```
// Not sure what this language is... JS?
let liferayEventFired;

while (!${liferayEventFired} && (maxIterations = "50")) {
    liferayEventFired = selenium.getEval("${javascript}");
    
    echo("${liferayEventFired}");

    if (!liferyEventFired) {
        Pause(locator1 = "100");
    }
}
```

Reply:
@wincent could you weigh in a bit here and in the rest of my comments if you have the chance?

Reply:
@jbalsas, everything on the JS side, you guys can change, which is what I was expecting. I just got it to work with what I had. All this while logic is going to go away when we move it into a *.function file though. Will send some changes soon.

Reply:
The main thing that sticks out to me is that the use targeting of `dialog-iframe-popup`, which doesn't seem to have any identifying power. For example, in `modules/apps/portal-workflow/portal-workflow-test/src/testFunctional/macros/Workflow.macro`, you can see we wait for it twice in a row:

```
		LiferayEvent.waitForLiferayEvent(
			attributeName = "src",
			attributeValue = "dialog-iframe-popup",
			eventName = "modalIframeLoaded");

		SelectFrame(locator1 = "WorkflowDefinition#DIALOG_BOX_BODY");

		LiferayEvent.waitForLiferayEvent(
			attributeName = "src",
			attributeValue = "dialog-iframe-popup",
			eventName = "modalIframeLoaded");
```

Now, I can't see the implementation of `waitForLiferayEvent` in this pull (I gather that went into Poshi via a separate update), so I don't know exactly what the semantics are. But if it just scans the list of events without `SelectFrame` having some intermediate effect, I would expect the second call to return immediately because the event is already in the list. I don't know if it maintains an internal pointer, for example, which would make it consider only subsequent events.

I am hesitant to say any more though because I don't fully understand how all this works and I don't want to add to the confusion that may already be present in our various threads.

Reply:
Yup, see https://github.com/john-co/liferay-portal/pull/293#discussion_r422134235 or my several different rants about this in the Slack channel...

Reply:
This, @wincent -> https://liferay.slack.com/archives/G013RPK014G/p1588937057195600

Reply:
> Why not something like:
> 
> ```js
> const event = LFR_JS_EVENTS_LOG.getLastEvent("${eventName}");
> 
> // It would be better to use the [] syntax to protect from attribute names with illegal characters for the dot notation.
> if (event['${attributeName}'] == '${attributeValue}') {
>     return JSON.stringify(event);
> }
> ```

The problem I see with this API is that for a use case like the iframe `src` checking that we're talking about here, it would require a perfect/complete string match. That means if we ever tweak the URL generation in ways that are not semantically significant (eg. if query string param order changes) then we'd wind up with a bunch of broken tests. I like the way the `includes` version allows us to make an assertion based on a specific aspect of the URL (even though we're not doing it right in this PR).

I don't want to hold up this long-running PR any further, but my wishlist for an iteration would be the ability to specify matches in different ways:

```
LiferayEvent.waitForLiferayEvent(
			attributeName = "src",
			attributeValue = "expect value to match this exactly",
			eventName = "modalIframeLoaded");

LiferayEvent.waitForLiferayEvent(
			attributeName = "src",
			attributeIncludes = "expect value to contain this",
			eventName = "modalIframeLoaded");

LiferayEvent.waitForLiferayEvent(
			attributeName = "src",
			attributeMatches = "expect value to match this regular expression pattern",
			eventName = "modalIframeLoaded");
```

Reply:
Many 👍 from me :)